### ISSUE TITLE
ServiceBalancerJob loop on daemon services with affinity filter

### ISSUE THREAD

**Issue by SpComb** [kontena/kontena#1895]
Running a `strategy: daemon` service with `affinity: label...` filters to only run on a subset of grid nodes results in a continous `ServiceBalancerJob: rebalancing service ....` loop. It seems like the `ServiceBalancerJob` `should_balance_stateless_service?` -> `all_instances_exist?` -> `desired_count_for_service` logic does not take the scheduling filters into account for daemon services. The rebalancer expects each daemon service to be running on all grid nodes, and thus attempts to deploy it. 

### `kontena node ls`
```
  Name                                                                   Version    Status     Initial    Labels
⊛ core-01                                                                1.1.2.rc1  online     1 / 1      test
⊛ core-02                                                                1.1.2.rc1  online     -          no-ingress-lb
```

### `kontena service show ingress-lb/lb`
```
development/ingress-lb/lb:
  created: 2017-02-22T10:41:22.509Z
  updated: 2017-02-22T10:41:22.509Z
  stack: development/ingress-lb
  state: running
  image: kontena/lb:latest
  revision: 1
  stack_revision: 1
  stateful: no
  scaling: 1
  strategy: daemon
  deploy_opts:
    min_health: 
    wait_for_port: 80
    interval: 604800
  dns: lb.ingress-lb.development.kontena.local
  affinity: 
    - label!=no-ingress-lb
  secrets: 
    - secret: LB_STATS_PASSWORD
      name: STATS_PASSWORD
      type: env
  env: 
    - KONTENA_LB_HEALTH_URI=/__health
  net: bridge
  ports:
    - 80:80/tcp
    - 443:443/tcp
  health check:
    protocol: http
    uri: /__health
    port: 80
    timeout: 10
    interval: 60
    initial_delay: 10
  health status:
    healthy: 1
    total: 1
  instances:
    ingress-lb.lb-1:
      rev: 2017-02-24 11:00:56 UTC
      service_rev: 1
      node: core-01
      dns: lb-1.ingress-lb.development.kontena.local
      ip: 10.81.128.66
      public ip: 82.181.164.219
      health: healthy (58s ago)
      status: running
      exit code: 0
```

### `kontena-server` logs
```
I, [2017-02-24T13:01:18.182670 #23219]  INFO -- ServiceBalancerJob: rebalancing service: development/ingress-lb/lb
I, [2017-02-24T13:01:18.230342 #23223]  INFO -- GridServiceDeployer: starting to deploy development/ingress-lb/lb
I, [2017-02-24T13:01:18.256393 #23223]  INFO -- GridServiceDeployer: deploying service instance development/ingress-lb/lb-1 to node core-01
I, [2017-02-24T13:01:18.256765 #23223]  INFO -- GridServiceDeployer: throttling service instance development/ingress-lb/lb deploy because of min_health limit (1 instances in-progress)
I, [2017-02-24T13:01:21.957685 #23223]  INFO -- GridServiceDeployer: service development/ingress-lb/lb has been deployed
I, [2017-02-24T13:01:34.194585 #23215]  INFO -- LeaderElectorJob: won election ♚
I, [2017-02-24T13:01:38.258360 #23215]  INFO -- ServiceBalancerJob: rebalancing service: development/ingress-lb/lb
I, [2017-02-24T13:01:38.715953 #23219]  INFO -- GridServiceDeployer: starting to deploy development/ingress-lb/lb
I, [2017-02-24T13:01:38.809646 #23219]  INFO -- GridServiceDeployer: deploying service instance development/ingress-lb/lb-1 to node core-01
I, [2017-02-24T13:01:38.810035 #23219]  INFO -- GridServiceDeployer: throttling service instance development/ingress-lb/lb deploy because of min_health limit (1 instances in-progress)
I, [2017-02-24T13:01:42.533291 #23219]  INFO -- GridServiceDeployer: service development/ingress-lb/lb has been deployed
I, [2017-02-24T13:01:58.205907 #23215]  INFO -- ServiceBalancerJob: rebalancing service: development/ingress-lb/lb
I, [2017-02-24T13:01:58.605193 #23219]  INFO -- GridServiceDeployer: starting to deploy development/ingress-lb/lb
I, [2017-02-24T13:01:58.672945 #23219]  INFO -- GridServiceDeployer: deploying service instance development/ingress-lb/lb-1 to node core-01
I, [2017-02-24T13:01:58.673284 #23219]  INFO -- GridServiceDeployer: throttling service instance development/ingress-lb/lb deploy because of min_health limit (1 instances in-progress)
I, [2017-02-24T13:02:02.375805 #23219]  INFO -- GridServiceDeployer: service development/ingress-lb/lb has been deployed
I, [2017-02-24T13:02:18.210269 #23215]  INFO -- ServiceBalancerJob: rebalancing service: development/ingress-lb/lb
I, [2017-02-24T13:02:18.513984 #23219]  INFO -- GridServiceDeployer: starting to deploy development/ingress-lb/lb
I, [2017-02-24T13:02:18.575474 #23219]  INFO -- GridServiceDeployer: deploying service instance development/ingress-lb/lb-1 to node core-01
I, [2017-02-24T13:02:18.575859 #23219]  INFO -- GridServiceDeployer: throttling service instance development/ingress-lb/lb deploy because of min_health limit (1 instances in-progress)
I, [2017-02-24T13:02:22.791369 #23219]  INFO -- GridServiceDeployer: service development/ingress-lb/lb has been deployed
I, [2017-02-24T13:02:32.206800 #23223]  INFO -- LeaderElectorJob: won election ♚
I, [2017-02-24T13:02:38.241463 #23223]  INFO -- ServiceBalancerJob: rebalancing service: development/ingress-lb/lb
I, [2017-02-24T13:02:38.896039 #23219]  INFO -- GridServiceDeployer: starting to deploy development/ingress-lb/lb
I, [2017-02-24T13:02:38.936099 #23219]  INFO -- GridServiceDeployer: deploying service instance development/ingress-lb/lb-1 to node core-01
I, [2017-02-24T13:02:38.936454 #23219]  INFO -- GridServiceDeployer: throttling service instance development/ingress-lb/lb deploy because of min_health limit (1 instances in-progress)
I, [2017-02-24T13:02:42.658394 #23219]  INFO -- GridServiceDeployer: service development/ingress-lb/lb has been deployed
I, [2017-02-24T13:02:58.130696 #23223]  INFO -- ServiceBalancerJob: rebalancing service: development/ingress-lb/lb
I, [2017-02-24T13:02:58.157493 #23215]  INFO -- GridServiceDeployer: starting to deploy development/ingress-lb/lb
I, [2017-02-24T13:02:58.197567 #23215]  INFO -- GridServiceDeployer: deploying service instance development/ingress-lb/lb-1 to node core-01
I, [2017-02-24T13:02:58.197940 #23215]  INFO -- GridServiceDeployer: throttling service instance development/ingress-lb/lb deploy because of min_health limit (1 instances in-progress)
I, [2017-02-24T13:03:02.365853 #23215]  INFO -- GridServiceDeployer: service development/ingress-lb/lb has been deployed
...
```

### `kontena-agent` logs
```
kontena-agent | D, [2017-02-24T11:04:38.739724 #1] DEBUG -- RpcServer: rpc request: Kontena::Rpc::ServicePodsApi#create [{"service_id"=>"58ad6ad2de3578692900007a", "service_name"=>"ingress-lb.lb", "service_revision"=>1, "updated_at"=>"2017-02-22 10:41:22 UTC", "instance_number"=>1, "image_name"=>"kontena/lb:latest", "image_credentials"=>nil, "deploy_rev"=>"2017-02-24 11:04:38 UTC", "stateful"=>false, "user"=>nil, "cmd"=>nil, "entrypoint"=>nil, "memory"=>nil, "memory_swap"=>nil, "cpu_shares"=>nil, "privileged"=>nil, "cap_add"=>[], "cap_drop"=>[], "devices"=>[], "ports"=>[{"ip"=>"0.0.0.0", "protocol"=>"tcp", "node_port"=>80, "container_port"=>80}, {"ip"=>"0.0.0.0", "protocol"=>"tcp", "node_port"=>443, "container_port"=>443}], "volumes"=>[], "volumes_from"=>[], "net"=>"bridge", "hostname"=>"lb-1", "domainname"=>"ingress-lb.development.kontena.local", "exposed"=>true, "log_driver"=>nil, "log_opts"=>{}, "pid"=>nil, "env"=>["KONTENA_LB_HEALTH_URI=/__health", "KONTENA_SERVICE_ID=58ad6ad2de3578692900007a", "KONTENA_SERVICE_NAME=lb", "KONTENA_GRID_NAME=development", "KONTENA_STACK_NAME=ingress-lb", "KONTENA_NODE_NAME=core-01", "KONTENA_SERVICE_INSTANCE_NUMBER=1"], "secrets"=>[{"name"=>"STATS_PASSWORD", "type"=>"env", "value"=>"OWNQkluHOPBS"}], "labels"=>{"io.kontena.container.id"=>"58b01346de35785ab3000007", "io.kontena.service.id"=>"58ad6ad2de3578692900007a", "io.kontena.service.name"=>"lb", "io.kontena.stack.name"=>"ingress-lb", "io.kontena.grid.name"=>"development", "io.kontena.health_check.uri"=>"/__health", "io.kontena.health_check.protocol"=>"http", "io.kontena.health_check.interval"=>"60", "io.kontena.health_check.timeout"=>"10", "io.kontena.health_check.initial_delay"=>"10", "io.kontena.health_check.port"=>"80"}, "hooks"=>[], "networks"=>[{"name"=>"kontena", "subnet"=>"10.81.0.0/16", "multicast"=>true, "internal"=>false}]}]
kontena-agent | I, [2017-02-24T11:04:38.745178 #1]  INFO -- Kontena::ServicePods::Creator: creating service: ingress-lb.lb-1
kontena-agent | I, [2017-02-24T11:04:38.745527 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulling image: kontena/lb:latest
kontena-agent | I, [2017-02-24T11:04:41.807099 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulled image: kontena/lb:latest
kontena-agent | I, [2017-02-24T11:04:41.926239 #1]  INFO -- Kontena::ServicePods::Creator: service is up-to-date: ingress-lb.lb-1
kontena-agent | D, [2017-02-24T11:04:42.384624 #1] DEBUG -- RpcServer: rpc request: Kontena::Rpc::AgentApi#port_open? ["10.81.128.66", 80]
kontena-agent | D, [2017-02-24T11:04:58.648268 #1] DEBUG -- RpcServer: rpc request: Kontena::Rpc::ServicePodsApi#create [{"service_id"=>"58ad6ad2de3578692900007a", "service_name"=>"ingress-lb.lb", "service_revision"=>1, "updated_at"=>"2017-02-22 10:41:22 UTC", "instance_number"=>1, "image_name"=>"kontena/lb:latest", "image_credentials"=>nil, "deploy_rev"=>"2017-02-24 11:04:58 UTC", "stateful"=>false, "user"=>nil, "cmd"=>nil, "entrypoint"=>nil, "memory"=>nil, "memory_swap"=>nil, "cpu_shares"=>nil, "privileged"=>nil, "cap_add"=>[], "cap_drop"=>[], "devices"=>[], "ports"=>[{"ip"=>"0.0.0.0", "protocol"=>"tcp", "node_port"=>80, "container_port"=>80}, {"ip"=>"0.0.0.0", "protocol"=>"tcp", "node_port"=>443, "container_port"=>443}], "volumes"=>[], "volumes_from"=>[], "net"=>"bridge", "hostname"=>"lb-1", "domainname"=>"ingress-lb.development.kontena.local", "exposed"=>true, "log_driver"=>nil, "log_opts"=>{}, "pid"=>nil, "env"=>["KONTENA_LB_HEALTH_URI=/__health", "KONTENA_SERVICE_ID=58ad6ad2de3578692900007a", "KONTENA_SERVICE_NAME=lb", "KONTENA_GRID_NAME=development", "KONTENA_STACK_NAME=ingress-lb", "KONTENA_NODE_NAME=core-01", "KONTENA_SERVICE_INSTANCE_NUMBER=1"], "secrets"=>[{"name"=>"STATS_PASSWORD", "type"=>"env", "value"=>"OWNQkluHOPBS"}], "labels"=>{"io.kontena.container.id"=>"58b0135ade35785abb00000a", "io.kontena.service.id"=>"58ad6ad2de3578692900007a", "io.kontena.service.name"=>"lb", "io.kontena.stack.name"=>"ingress-lb", "io.kontena.grid.name"=>"development", "io.kontena.health_check.uri"=>"/__health", "io.kontena.health_check.protocol"=>"http", "io.kontena.health_check.interval"=>"60", "io.kontena.health_check.timeout"=>"10", "io.kontena.health_check.initial_delay"=>"10", "io.kontena.health_check.port"=>"80"}, "hooks"=>[], "networks"=>[{"name"=>"kontena", "subnet"=>"10.81.0.0/16", "multicast"=>true, "internal"=>false}]}]
kontena-agent | I, [2017-02-24T11:04:58.655937 #1]  INFO -- Kontena::ServicePods::Creator: creating service: ingress-lb.lb-1
kontena-agent | I, [2017-02-24T11:04:58.658910 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulling image: kontena/lb:latest
kontena-agent | I, [2017-02-24T11:05:01.675429 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulled image: kontena/lb:latest
kontena-agent | I, [2017-02-24T11:05:01.768975 #1]  INFO -- Kontena::ServicePods::Creator: service is up-to-date: ingress-lb.lb-1
kontena-agent | D, [2017-02-24T11:05:02.060890 #1] DEBUG -- Workers::StatsWorker: starting collection
kontena-agent | D, [2017-02-24T11:05:02.265036 #1] DEBUG -- RpcServer: rpc request: Kontena::Rpc::AgentApi#port_open? ["10.81.128.66", 80]
```

**Comment by cooperaj** [comment]
![CPU issue](https://www.dropbox.com/s/kmsug4cmcrijqkg/srv10.png?dl=0)

As you can see above this looping appears to be causing some issues :)

**Comment by SpComb** [comment]
The rebalancer loop happens every 20 seconds, so the pattern there would be be on a much shorter timescale.

Interesting to see the constant level of CPU utilization increase over time, though. Is it the `kontena-server-api` Docker container -> `ruby`/`puma` process that's eating CPU, or the `kontena-server-mongo` Docker container -> `mongodb` process?

**Comment by cooperaj** [comment]
It appears to be the server-api container. The mongo container maintains a fairly consistent cpu usage. (and was not the one restarted in the above image)

**Comment by jakolehm** [comment]
#1873 should fix this

**Comment by colinmollenhour** [comment]
Any idea when 1.2.0 will be ready for production or any chance a fix could be applied to 1.1? This issue is making my Kontena stacks very unreliable (service instances deploying to all nodes and then later disappearing) and is the only issue preventing me from being comfortable with completing a migration to Kontena.

**Comment by colinmollenhour** [comment]
Or is there at least a temporary fix besides removing the entire stack and re-deploying? Right now for example I have a service with only 2 instances running when there should be 4 and none of the non-destructive commands I've tried can revive the other 2. E.g. 'kontena service restart', 'kontena service deploy --force', restarting the Kontena master API container, etc...

Log example:
```
I, [2017-03-22T18:00:29.513860 #6]  INFO -- GridServiceDeployer: starting to deploy primary/redis-cache/lb
I, [2017-03-22T18:00:33.108473 #6]  INFO -- GridServiceDeployer: deploying service instance primary/redis-cache/lb-1 to node app2
I, [2017-03-22T18:00:33.115329 #6]  INFO -- GridServiceDeployer: throttling service instance primary/redis-cache/lb deploy because of min_health limit (1 instances in-progress)
I, [2017-03-22T18:00:39.583310 #6]  INFO -- GridServiceDeployer: deploying service instance primary/redis-cache/lb-2 to node app1
I, [2017-03-22T18:00:39.586034 #6]  INFO -- GridServiceDeployer: throttling service instance primary/redis-cache/lb deploy because of min_health limit (1 instances in-progress)
```

The problem is that there are two other nodes besides app1 and app2 that never get the esrvice deployed to them because of some issue with this service getting stuck on app1 and app2. If I deploy the stack fresh it works and then some number of days later at some point it breaks ostensibly due to this service deploy thrashing.

**Comment by colinmollenhour** [comment]
I've been digging in deeper and found that the error is that the 'wait_for_port' option is timing out **even though the service is running and accepting connections on that port**. I started using "nc -z <ip> <port>" to confirm if it should really be timing out and found something very interesting.. On the service that is broken if I use "nc -z" against the bridge IP it works and if I use it against the weave IP it doesn't. For a different service with an almost identical image and service config the "nc -z" test works against BOTH the bridge IP and the weave IP.. So I'm guessing that "container.ip_address" on [this line](https://github.com/kontena/kontena/blob/v1.1/server/app/services/grid_service_instance_deployer.rb#L120) is the weave IP and not the Docker bridge IP? However, I have no idea why the host and agent would not be able to connect to the new container over the weave IP but it would over the docker IP; that is beyond my knowledge of weave/kontena..

Here is proof from the kontena-agent container:

```
/app # getent hosts lb-1.galera.primary.kontena.local
10.81.128.50      lb-1.galera.primary.kontena.local  lb-1.galera.primary.kontena.local
/app # getent hosts lb-1.redis-cache.primary.kontena.local
10.81.128.60      lb-1.redis-cache.primary.kontena.local  lb-1.redis-cache.primary.kontena.local
/app # nc -z 10.81.128.50 1936 && echo yes
yes
/app # nc -z 10.81.128.60 1936 && echo yes
/app # nc -z 172.17.0.2 1936 && echo yes
yes
```

The next to last line is the kontena network IP, the last line is the Docker bridge IP of the same container.

**Comment by SpComb** [comment]
> Any idea when 1.2.0 will be ready for production or any chance a fix could be applied to 1.1? This issue is making my Kontena stacks very unreliable (service instances deploying to all nodes and then later disappearing) and is the only issue preventing me from being comfortable with completing a migration to Kontena.

This issue is about the server triggering unnecessary re-deploys, which should be harmless, as the services are already deployed and up to date. If these extra deploys are then in turn triggering other bugs that make your services unstable, then that's a separate issue.

> I've been digging in deeper and found that the error is that the 'wait_for_port' option is timing out even though the service is running and accepting connections on that port.

Let's figure this one out over in #1862

**Comment by SpComb** [comment]
Fixed by #1873

```
D, [2017-03-27T13:44:24.728149 #8740] DEBUG -- GridScheduler: service does not need rescheduling: development/null/redis-daemon
```
